x-pack/plugins/enterprise_search/server/utils/create_ml_inference_pipeline.ts

```diff
@@ -0,0 +1,134 @@
+
+import { formatMlPipelineBody } from './create_pipeline_definitions';
+
+/**
+ * Details of a created or updated pipeline.
+ */
+export interface ModifiedPipeline {
+  id: string;
+  created?: boolean;
+  updated?: boolean;
+}
+
+/**
+ * Creates a Machine Learning Inference pipeline with the given settings, if it doesn't exist yet.
+ * @param pipelineName pipeline name set by the user.
+ * @param modelId model ID selected by the user.
+ * @param sourceField The document field that model will read.
+ * @param destinationField The document field that the model will write to.
+ * @param esClient the Elasticsearch Client to use when retrieving pipeline and model details.
+ */
+export const createMlInferencePipeline = async (
+  pipelineName: string,
+  modelId: string,
+  sourceField: string,
+  destinationField: string,
+  esClient: ElasticsearchClient
+): Promise<ModifiedPipeline> => {
+  const inferencePipelineGeneratedName = `ml-inference-${pipelineName}`;
+
+  // Check that a pipeline with the same name doesn't already exist
+  let pipelineByName: IngestGetPipelineResponse | undefined;
+  try {
+    pipelineByName = await esClient.ingest.getPipeline({
+      id: inferencePipelineGeneratedName,
+    });
+  } catch (error) {
+    // Silently swallow error
+  }
+  if (pipelineByName?.[inferencePipelineGeneratedName]) {
+    throw new Error(ErrorCode.PIPELINE_ALREADY_EXISTS);
+  }
+
+  // Generate pipeline with default processors
+  const mlInferencePipeline = await formatMlPipelineBody(
+    modelId,
+    sourceField,
+    destinationField,
+    esClient
+  );
+
+  await esClient.ingest.putPipeline({
+    id: inferencePipelineGeneratedName,
+    ...mlInferencePipeline,
+  });
+
+  return Promise.resolve({
+    id: inferencePipelineGeneratedName,
+    created: true,
+  });
+};
+
+/**
+ * Adds the supplied a Machine Learning Inference pipeline reference to the "parent" ML Inference
+ * pipeline that is associated with the index.
+ * @param indexName name of the index this pipeline corresponds to.
+ * @param pipelineName name of the ML Inference pipeline to add.
+ * @param esClient the Elasticsearch Client to use when retrieving pipeline details.
+ */
+export const addSubPipelineToIndexSpecificMlPipeline = async (
+  indexName: string,
+  pipelineName: string,
+  esClient: ElasticsearchClient
+): Promise<ModifiedPipeline> => {
+  const parentPipelineId = `${indexName}@ml-inference`;
+
+  // Fetch the parent pipeline
+  let parentPipeline: IngestPipeline | undefined;
+  try {
+    const pipelineResponse = await esClient.ingest.getPipeline({
+      id: parentPipelineId,
+    });
+    parentPipeline = pipelineResponse[parentPipelineId];
+  } catch (error) {
+    // Swallow error; in this case the next step will return
+  }
+
+  // Verify the parent pipeline exists with a processors array
+  if (!parentPipeline?.processors) {
+    return Promise.resolve({
+      id: parentPipelineId,
+      updated: false,
+    });
```

Comment:
Gracefully handling the case when the `{index-name}@ml-inference` parent pipeline doesn't exist. (We still create the sub-pipeline for inference, but it won't be referenced in the parent.)

This should not happen though because adding an ML inference processor is only allowed if the "ML inference" checkbox is checked on the UI, which creates the `{index-name}@ml-inference` pipeline.